### PR TITLE
Issue 3312 - Additional real time events

### LINKED ISSUE #3312: Additional real time events
### Description
Extra v5 real time events to support the implementation of https://github.com/3drepo/3drepo.io/issues/2716

##### New Revision (Container)
scope: Container/Federation events
event name: `containerNewRevision`
triggered when: a new revision has been processed successfully.
Data format: 
````js
{
  teamspace: "teamspace",
  project: "projectId",
  model: "containerId",
  data: {
    _id: "revId",
    tag: "tag",
    timestamp: 123, // epoch ts
    owner: "abc"
  }
}
````
##### New Revision (Federation)
scope: Container/Federation events
event name: `federationNewRevision`
triggered when: a new revision has been processed successfully.
Data format: Same as `containerNewRevision`

##### Revision Updated (Container)
scope: Container/Federation events
event name: `containerRevisionUpdate`
triggered when: some data has changed within a revision
Data format: 
````js
{
  teamspace: "teamspace",
  project: "projectId",
  model: "containerId",
  data: {
    _id: "revId",
    void: true/false // This is the only field that can be changed in revision atm
  }
}
````

##### Delete Container
scope: Container/Federation events
event name: `containerRemoved`
triggered when: A container has been removed from the project
Data format: 
````js
{
  teamspace: "teamspace",
  project: "projectId",
  model: "containerId"
  data: {}
}
````

##### Delete Federation
scope: Container/Federation events
event name: `federationRemoved`
triggered when: A federation has been removed from the project
Data format: 
````js
{
  teamspace: "teamspace",
  project: "projectId",
  model: "containerId"
  data: {}
}
````

##### New Container
scope: Project event
event name: `newContainer`
triggered when: A new container has been added to the project
Data format: 
````js
{
  teamspace: "teamspace",
  project: "projectId",
  data: {
    _id: "container id",
    code: "container code",
    category: "category" ,
  }
}
````

##### New Federation
scope: Project event
event name: `newFederation`
triggered when: A new federationhas been added to the project
Data format: 
````js
{
  teamspace: "teamspace",
  project: "projectId",
  data: {
    _id: "container id",
    code: "container code",
    description: "abc"
  }
}
````

### Goals
- Real time events should trigger on the actions specified

### Tasks
- [x] Implement the events
- [ ] Add tests
- [ ] Add documentation to `backend/src/v5/services/chat/README.md`

### PR DESCRIPTION
**Pull request by ChristosTsiotsias**
This fixes #3312 

#### Description

The following real time events have been added:

New Revision (Container)
New Revision (Federation)
Revision Updated (Container)
Delete Container
Delete Federation
New Container
New Federation

Documentation for each event has been added

#### Test cases

A real time event should be triggered upon each action
The events documentation should be updated to include all new events

### CODE REVIEW

File: backend/src/v5/models/modelSettings.js [hunk 1]
```diff
@@ -51,8 +57,6 @@ Models.deleteModel = async (ts, model) => {
 	if (deletedCount === 0) {
 		throw templates.modelNotFound;
 	}
-
```
Comment (carmenfan) [inline backend/src/v5/models/modelSettings.js:54]: not fond of how we're publishing the model related events here apart from this one.

Can do we a findOneAndDelete instead so you can figure out if it's a federation and trigger the event here (assuming this is the reason why you moved it out)

File: backend/src/v5/services/eventsListener/components/modelEvents.js [hunk 2]
```diff
@@ -38,6 +39,12 @@ const queueTasksCompleted = async ({
 	try {
 		const { _id: projectId } = await findProjectByModelId(teamspace, model, { _id: 1 });
 		await newRevisionProcessed(teamspace, UUIDToString(projectId), model, corId, value, user, containers);
+
+		const { tag, author, timestamp } = await getRevisionByIdOrTag(teamspace, model, corId,
```
Comment (carmenfan) [inline backend/src/v5/services/eventsListener/components/modelEvents.js:43]: Not the best place i don't think.

We might be processing a failed model import, in which case you're still attempting to trigger a new revision event here.

I think its' best if `newRevisionProcessed ` trigger a` NEW_REVISION` event upon model processing succeeded, and we catch it to create the chat message, it also knows whether it's a container so you don't have to do `getModelById`

File: backend/src/v5/services/eventsListener/components/projectEvents.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+const { EVENTS: chatEvents } = require('../../chat/chat.constants');
+const { createProjectMessage } = require('../../chat');
+const { events } = require('../../eventsManager/eventsManager.constants');
+const { logger } = require('../../../utils/logger');
+const { subscribe } = require('../../eventsManager/eventsManager');
+
+const modelAdded = async ({ teamspace, project, model, data, isFederation }) => {
```
Comment (carmenfan) [inline backend/src/v5/services/eventsListener/components/projectEvents.js:24]: wrong place

This is a model event.

(even though in the chat service, it's broadcasted in the project room)

There's a difference between the event managers within the backend and the chat service events, they're 2 different things

File: backend/src/v5/services/eventsManager/eventsManager.constants.js [hunk 4]
```diff
@@ -23,7 +23,7 @@ const eventList = [
 	// Model settngs
 	'NEW_MODEL', 'DELETE_MODEL', 'MODEL_SETTINGS_UPDATE',
 	// Model import (including federation and toy project)
-	'MODEL_IMPORT_FINISHED',
+	'MODEL_IMPORT_FINISHED', 'REVISION_UPDATED',
```
Comment (carmenfan) [inline backend/src/v5/services/eventsManager/eventsManager.constants.js:26]: Revision update isn't part of import, just put it under revisions

**Review by carmenfan — CHANGES_REQUESTED** [review]
Missing code to allow a user to join a project room, as it stands it's impossible for anyone to subscribe to a project event.

https://github.com/3drepo/3drepo.io/blob/195c67e839d17873827b926cf4d02beca42d7f26/backend/src/v5/services/chat/socketsManager.js#L40

There should also be e2e tests for all those chat events you've created (if you've done them you'd realise there's no way to join a project room!), see here for examples: https://github.com/3drepo/3drepo.io/tree/195c67e839d17873827b926cf4d02beca42d7f26/backend/tests/v5/e2e/services/chat/modelEvents

File: backend/tests/v5/e2e/services/chat/modelEvents/modelUploads.test.js [hunk 5]
```diff
@@ -221,6 +225,103 @@ const queueFinishedTest = () => {
...
+					tag: federationRevision.tag,
+					timestamp: results.data.timestamp,
+				} }));
+
+			socket.close();
+		});
+	});
+};
+
+const revisionUpdateTest = () => {
```
Comment (carmenfan) [inline backend/tests/v5/e2e/services/chat/modelEvents/modelUploads.test.js:282]: as discussed this should go into modeRevisions file separately

File: backend/tests/v5/e2e/services/chat/modelEvents/modelUploads.test.js [hunk 6]
```diff
@@ -221,6 +225,103 @@ const queueFinishedTest = () => {
 	});
 };
 
+const revisionAddedTest = () => {
```
Comment (carmenfan) [inline backend/tests/v5/e2e/services/chat/modelEvents/modelUploads.test.js:228]: as discussed, you shouldn't need an additional test for this? `queueFinishedTest` essentially is the same situation

File: backend/tests/v5/unit/models/modelSettings.test.js [hunk 7]
```diff
@@ -207,51 +207,110 @@ const testGetFederations = () => {
...
+			expect(EventsManager.publish).toHaveBeenCalledTimes(1);
+			expect(EventsManager.publish).toHaveBeenCalledWith(events.NEW_MODEL, { teamspace,
+				project,
+				model: fn.mock.calls[0][2]._id,
+				data: { code: data.properties.code, category: data.type },
+				isFederation: false });
+		});
+	});
+
+	describe('Add model', () => {
```
Comment (carmenfan) [inline backend/tests/v5/unit/models/modelSettings.test.js:236]: redundant, you already created a `Add model` test group

File: backend/tests/v5/unit/models/modelSettings.test.js [hunk 8]
```diff
@@ -207,51 +207,110 @@ const testGetFederations = () => {
...
+			const project = generateRandomString();
+			const model = generateRandomString();
+			await expect(Model.deleteModel(teamspace, project, model))
 				.rejects.toEqual(templates.modelNotFound);
-			expect(fn.mock.calls.length).toBe(1);
-			expect(fn.mock.calls[0][0]).toEqual(teamspace);
-			expect(fn.mock.calls[0][1]).toEqual('settings');
-			expect(fn.mock.calls[0][2]).toEqual({ _id: modelId });
+			expect(fn).toHaveBeenCalledTimes(1);
+			expect(fn).toHaveBeenCalledWith(teamspace, 'settings', { _id: model }, { federate: 1 });
```
Comment (carmenfan) [inline backend/tests/v5/unit/models/modelSettings.test.js:313]: should add a line to ensure `EventsManager.publish` has not been called

File: backend/tests/v5/unit/services/eventsListener/eventsListener.test.js [hunk 9]
```diff
@@ -51,30 +55,66 @@ const eventTriggeredPromise = (event) => new Promise(
 
 const testModelEventsListener = () => {
 	describe('Model Events', () => {
-		test(`Should trigger ModelStatusUpdate if there is a ${events.QUEUED_TASK_UPDATE}`, async () => {
+		test(`Should trigger queueStatusUpdate if there is a ${events.QUEUED_TASK_UPDATE}`, async () => {
```
Comment (carmenfan) [inline backend/tests/v5/unit/services/eventsListener/eventsListener.test.js:58]:
```suggestion
		test(`Should trigger updateModelStatusif there is a ${events.QUEUED_TASK_UPDATE}`, async () => {
```

File: backend/tests/v5/unit/services/eventsListener/eventsListener.test.js [hunk 10]
```diff
@@ -51,30 +55,66 @@ const eventTriggeredPromise = (event) => new Promise(
...
+			expect(ProjectSettings.findProjectByModelId).toHaveBeenCalledWith(data.teamspace, data.model, { _id: 1 });
 			expect(ModelSettings.newRevisionProcessed).toHaveBeenCalledTimes(1);
-			expect(ModelSettings.newRevisionProcessed).toHaveBeenCalledWith(
-				data.teamspace, project, data.model, data.corId, data.value, data.user, undefined,
-			);
+			expect(ModelSettings.newRevisionProcessed).toHaveBeenCalledWith(data.teamspace, project, data.model,
+				data.corId, data.value, data.user, data.containers);
+		});
+
+		test(`Should trigger queueTasksCompleted if there is a ${events.QUEUED_TASK_COMPLETED} (federation)`, async () => {
```
Comment (carmenfan) [inline backend/tests/v5/unit/services/eventsListener/eventsListener.test.js:97]: same here

File: backend/tests/v5/unit/services/eventsListener/eventsListener.test.js [hunk 11]
```diff
@@ -146,6 +186,276 @@ const testModelEventsListener = () => {
...
+				chatEvents.CONTAINER_REVISION_UPDATE,
+				{ ...data.data, _id: UUIDToString(data.data._id) },
+				data.teamspace,
+				data.project,
+				data.model,
+				undefined,
+			);
+		});
+
+		test(`Should trigger modelAdded if there is a ${events.NEW_MODEL} (federation)`, async () => {
```
Comment (carmenfan) [inline backend/tests/v5/unit/services/eventsListener/eventsListener.test.js:235]: better description

File: backend/tests/v5/unit/services/eventsListener/eventsListener.test.js [hunk 12]
```diff
@@ -146,6 +186,276 @@ const testModelEventsListener = () => {
...
+				chatEvents.CONTAINER_REVISION_UPDATE,
+				{ ...data.data, _id: UUIDToString(data.data._id) },
+				data.teamspace,
+				data.project,
+				data.model,
+				undefined,
+			);
+		});
+
+		test(`Should fail gracefully on error if there is a ${events.REVISION_UPDATED}`, async () => {
```
[the remark this answers is not in the document]
Comment (ChristosTsiotsias) [reply backend/tests/v5/unit/services/eventsListener/eventsListener.test.js:212]: Isnt the modelSettingsUpdated unlikely to error as well then ? Both are just calling createModelMessage

image

[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/tests/v5/unit/services/eventsListener/eventsListener.test.js:212]: yep that's true. it also shouldn't be wrapped and error handling should be done on the other side

[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/tests/v5/unit/services/eventsListener/eventsListener.test.js:212]: yes, let's do that

**Review by carmenfan — APPROVED** [review]

**Pull request merged**